### DISCUSSION TITLE
Codespaces 502 Bad Gateway on docfx serve

### DISCUSSION THREAD

**Discussion by SeanKilleen** [community/community#4070] [Codespaces] [answered]
Hi all,

I am working to add GitHub Codespaces to the NUnit docs repository for document editing. We're really excited about the potential of Codespaces for our use case!

## The Current Ingredients

* PR to add codespaces, which has our configuration files and containers: https://github.com/nunit/docs/pull/563
* The source of the Dockerfile we're referencing from our Codespaces Dockerfile: https://github.com/nunit/docfx-action/blob/master/Dockerfile

## Reproduction

* Open the codespace on the `github-codespaces` branch (can be done from the GitHub UI on the PR link above)
* Open the terminal
* `cd docs`
* `docfx build` - builds successfully, as expected
* `docfx serve` - indicates it is serving on port `8080`, as expected
* Follow the link from the terminal, which opens the port `8080` in a forwarded tab

## Expected Behavior

* See a listing of directories at the root of the site, including `_site`, which is how docfx typically serves them
* See the site at the `/_site` subdirectory which is where docfx outputs its files.

## Actual Behavior

See a 502 bad gateway error at the initial URL, and also when attempting with `/_site` appended.

## Other Things I've tried

* I've pulled this branch in VS Code on my machine and opened it in this container. I was able to see the site served as expected on both the forwarded `8080` port root and with `/_site` appended.

____

I assume this has to be something I'm doing. I'm looking into it to rule out that we're the cause of the issue since I'm new to Codespaces, but if anyone has a tip I'd appreciate it.

**Comment by SeanKilleen** [comment]
Cross-posted on SO in case it's beneficial that it be answered there: https://stackoverflow.com/questions/67791687/github-codespaces-502-gateway-error-serving-a-docfx-site

**Comment by 2percentsilk** [comment]
Hey @SeanKilleen, thanks for filing! I'll see if I can take a look at this today :)

**Reply by SeanKilleen** [reply]
Hi @2percentsilk -- by any chance have you gotten to look into this yet?

**Comment by SeanKilleen** [comment]
Some more information as I continue to attempt to troubleshoot this. 

I see that -- in the terminal inside the vs code window in my browser - the application is running:

> ![image](https://user-images.githubusercontent.com/2148318/130497801-1c4c3909-217e-4834-ae91-cc22aa324ef5.png)

But, if I go to the `Ports` tab, it seems that it is not detected as running:

> ![image](https://user-images.githubusercontent.com/2148318/130497983-6dd5dd4f-1516-4140-9b3c-087f9b21c5b5.png)

I'm trying to find resources to troubleshoot this but am having difficulty getting the search terms correct. 

@2percentsilk any chance this would help you dive in? I know you'd mentioned back in June that you would try to look into this. (no worries that it hasn't happened yet; I understand this isn't priority and you've got a lot going on.)

**Comment by SeanKilleen** [comment]
Interestingly enough -- when I instead run `docfx serve -n 127.0.0.1 -p 8080`:

> ![image](https://user-images.githubusercontent.com/2148318/130500968-160b09fc-80ab-4832-8c05-81df206d4ba5.png)

I do see the port show up in the port forwarding: 

> ![image](https://user-images.githubusercontent.com/2148318/130501016-19b05443-455c-4710-8b34-ab45e79a5bc4.png)

Unfortunately, I still don't see the expected result -- I now see an "invalid host" message.

> ![image](https://user-images.githubusercontent.com/2148318/130501096-414e8451-eff6-474b-a63d-1e22df8d9ffc.png)

This may be more related to Mono. I'll try to do some looking.

**Comment by SeanKilleen** [comment, marked as answer]
I figured it out! Thanks to this comment: https://github.com/dotnet/docfx/issues/3615#issuecomment-451664865

In this case, because docfx is using `WebApp.Start()`, the correct syntax inside of the container is:

`docfx serve _site -n "*"`

That `"*"` is what makes the difference here and allows mono to bind appropriately for the forwarding to work.
